src/wmi.c

```diff
@@ -0,0 +1,379 @@
+  char *plugin_instance_s = NULL;
+
+  if (instances_from == NULL)
+    return ssnprintf_alloc(instance_prefix);
+
+  if (wmi_result_get_value(wmi_result, instances_from,
+                           &plugin_instance_base_v) != 0) {
+    log_err("failed to read field '%s'", instances_from);
+    return "";
+  }
+  plugin_instance_base_s = variant_get_string(&plugin_instance_base_v);
+  if (plugin_instance_base_s == NULL) {
+    log_err("failed to convert plugin_instance to string");
+    return "";
+  }
+  plugin_instance_s =
+      ssnprintf_alloc("%s%s", instance_prefix, plugin_instance_s);
+  VariantClear(&plugin_instance_base_v);
+  free(plugin_instance_base_s);
+  return plugin_instance_s;
+}
+
+static void submit(const char *type, const char *type_instance,
+                   const char *plugin_instance, value_t *values,
+                   size_t values_len) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  vl.values = values;
+  vl.values_len = values_len;
+
+  sstrncpy(vl.host, hostname_g, sizeof(vl.host));
+  sstrncpy(vl.plugin, "wmi", sizeof(vl.plugin));
+
+  sstrncpy(vl.plugin_instance, plugin_instance, sizeof(vl.plugin_instance));
+  sstrncpy(vl.type, type, sizeof(vl.type));
+  if (type_instance != NULL)
+    sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+
+  plugin_dispatch_values(&vl);
+}
+
+static void gauge_submit(const char *type, const char *type_instance,
+                         const char *plugin_instance, gauge_t value) {
+  submit(type, type_instance, plugin_instance, &(value_t){.gauge = value}, 1);
+}
+
+static void derive_submit(const char *type, const char *type_instance,
+                          const char *plugin_instance, derive_t value) {
+  submit(type, type_instance, plugin_instance, &(value_t){.derive = value}, 1);
+}
+
+static void absolute_submit(const char *type, const char *type_instance,
+                            const char *plugin_instance, absolute_t value) {
+  submit(type, type_instance, plugin_instance, &(value_t){.absolute = value},
+         1);
+}
+
+static void counter_submit(const char *type, const char *type_instance,
+                           const char *plugin_instance, counter_t value) {
+  submit(type, type_instance, plugin_instance, &(value_t){.counter = value}, 1);
+}
+
+static int wmi_exec_query(wmi_query_t *q) {
+  wmi_result_list_t *results;
+  results = wmi_query(wmi, q->statement);
+
+  if (results->count == 0) {
+    log_warn("no results for query %s.", q->statement);
+    wmi_result_list_release(results);
+    return 0;
+  }
+
+  wmi_result_t *result;
+  while ((result = wmi_get_next_result(results))) {
+
+    wmi_metric_list_t *mn;
+    for (mn = q->metrics; mn != NULL; mn = mn->next) {
+      VARIANT value_v;
+      char *plugin_instance_s = NULL;
+      const data_set_t *ds = NULL;
+      wmi_metric_t *m = mn->node;
+
+      ds = plugin_get_ds(m->type);
+      if (ds->ds_num != 1) {
+        log_err("data set for metric type '%s' has %" PRIsz
```

Comment:
This error would be logged on each invocation, spamming the logs. Consider using the complaint mechanism from https://github.com/collectd/collectd/blob/master/src/daemon/utils_complain.h

Reply:
Maybe we can move `plugin_get_ds` to some early stage here?

Reply:
Seems to work, although it's hard to tell if it's logging less than it was previously.

Reply:
The `m->type` value is taken from plugin configuration, if I understand things correctly.
So, you can do this check in init stage. Example: `src/tail_csv.c`,  `tcsv_init()`.

Hope this helps.